framework/include/cppmicroservices/detail/TrackedBundle.h

```diff
@@ -0,0 +1,125 @@
+
+#ifndef CPPMICROSERVICES_TRACKEDBUNDLE_H
+#define CPPMICROSERVICES_TRACKEDBUNDLE_H
+
+#include "cppmicroservices/BundleEvent.h"
+#include "cppmicroservices/detail/BundleAbstractTracked.h"
+#include "cppmicroservices/detail/TrackedBundleListener.h"
+
+#include "cppmicroservices/detail/CounterLatch.h"
+#include "cppmicroservices/detail/ScopeGuard.h"
+
+namespace cppmicroservices {
+
+namespace detail {
+
+/**
+ * This class is not intended to be used directly. It is exported to support
+ * the CppMicroServices bundle system.
+ */
+template<class TTT>
+class TrackedBundle
+  : public TrackedBundleListener
+  , public BundleAbstractTracked<Bundle, TTT, BundleEvent>
+{
+
+public:
+  using T = typename TTT::TrackedType;
+  using TrackedParamType = typename TTT::TrackedParamType;
+
+  TrackedBundle(BundleTracker<T>* bundleTracker,
+                BundleTrackerCustomizer<T>* customizer);
+
+  /**
+   * Method connected to bundle events for the
+   * <code>BundleTracker</code> class. This method must NOT be
+   * synchronized to avoid deadlock potential.
+   *
+   * @param event <code>BundleEvent</code> object from the framework.
+   */
+  void BundleChanged(const BundleEvent& event) override;
+
+  void WaitOnCustomizersToFinish();
+
+private:
+  using Superclass = BundleAbstractTracked<Bundle, TTT, BundleEvent>;
+
+  BundleTracker<T>* bundleTracker;
+  BundleTrackerCustomizer<T>* customizer;
+
+  CounterLatch latch;
+
+  /**
+   * Increment the tracking count and tell the tracker there was a
+   * modification.
+   *
+   * @GuardedBy this
+   */
+  void Modified() override;
+
+  /**
+   * Call the specific customizer adding method. This method must not be
+   * called while synchronized on this object.
+   *
+   * @param bundle Bundle to be tracked.
+   * @param related Action related object.
+   * @return Customized object for the tracked bundle or <code>null</code>
+   *         if the bundle is not to be tracked.
+   */
+  std::optional<TrackedParamType> CustomizerAdding(
+    Bundle bundle,
+    const BundleEvent& related) override;
+
+  /**
+   * Call the specific customizer modified method. This method must not be
+   * called while synchronized on this object.
+   *
+   * @param bundle Tracked bundle.
+   * @param related Action related object.
+   * @param object Customized object for the tracked bundle.
+   */
+  void CustomizerModified(Bundle bundle,
+                          const BundleEvent& related,
+                          const TrackedParamType& object) override;
+
+  /**
+   * Call the specific customizer removed method. This method must not be
+   * called while synchronized on this object.
+   *
+   * @param bundle Tracked bundle.
+   * @param related Action related object.
+   * @param object Customized object for the tracked bundle.
+   */
+  void CustomizerRemoved(Bundle bundle,
+                         const BundleEvent& related,
+                         const TrackedParamType& object) override;
+};
```

Comment:
Comments here should be explicit about when these callbacks are called (i.e., what it means for a Bundle to be Added, Modified, or Removed).

Reply:
I've cleaned up the behavioral explanations in the BundleTracker and the BundleTrackerCustomizer for the callbacks. Since these methods just wrap the callbacks, I will add a @see tag to the doxygen linking to the API to describe the calling conditions